Source/Core/Common/OpenXR.cpp

```diff
@@ -0,0 +1,714 @@
+  layer.layerFlags = 0;
+  layer.space = m_local_space;
+  layer.viewCount = VIEW_COUNT;
+  layer.views = projection_views;
+
+  const XrCompositionLayerBaseHeader* const layers[] = {
+      reinterpret_cast<XrCompositionLayerBaseHeader*>(&layer)};
+
+  XrFrameEndInfo frame_end_info{XR_TYPE_FRAME_END_INFO};
+  frame_end_info.displayTime = m_display_time;
+  frame_end_info.environmentBlendMode = XR_ENVIRONMENT_BLEND_MODE_OPAQUE;
+  frame_end_info.layerCount = uint32_t(std::size(layers));
+  frame_end_info.layers = layers;
+
+  XrResult result = xrEndFrame(m_session, &frame_end_info);
+
+  return XR_SUCCEEDED(result);
+}
+
+bool Session::CreateSwapchain(const std::vector<int64_t>& supported_formats)
+{
+  uint32_t view_count = VIEW_COUNT;
+  std::vector<XrViewConfigurationView> config_views(view_count, {XR_TYPE_VIEW_CONFIGURATION_VIEW});
+  XrResult result = xrEnumerateViewConfigurationViews(s_instance, s_system_id, VIEW_CONFIG_TYPE,
+                                                      view_count, &view_count, config_views.data());
+
+  if (XR_FAILED(result) || view_count != VIEW_COUNT)
+  {
+    ERROR_LOG(VIDEO, "OpenXR: xrEnumerateViewConfigurationViews: %d (%d)", result, view_count);
+    return false;
+  }
+
+  uint32_t swapchain_format_count = 0;
+  xrEnumerateSwapchainFormats(m_session, 0, &swapchain_format_count, nullptr);
+
+  std::vector<int64_t> swapchain_formats(swapchain_format_count);
+  result = xrEnumerateSwapchainFormats(m_session, swapchain_format_count, &swapchain_format_count,
+                                       swapchain_formats.data());
+
+  if (XR_FAILED(result))
+  {
+    ERROR_LOG(VIDEO, "OpenXR: xrEnumerateSwapchainFormats: %d", result);
+    return false;
+  }
+
+  // Require left/right views have identical sizes
+  if (config_views[0].recommendedImageRectWidth != config_views[1].recommendedImageRectWidth ||
+      config_views[0].recommendedImageRectHeight != config_views[1].recommendedImageRectHeight ||
+      config_views[0].recommendedSwapchainSampleCount !=
+          config_views[1].recommendedSwapchainSampleCount)
+  {
+    ERROR_LOG(VIDEO, "OpenXR: Recommended left/right swapchain image sizes differ.");
+    return false;
+  }
+
+  auto& selected_config_view = config_views[0];
+  m_swapchain_size = {int32_t(selected_config_view.recommendedImageRectWidth),
+                      int32_t(selected_config_view.recommendedImageRectHeight)};
+
+  const auto format = std::find_first_of(swapchain_formats.begin(), swapchain_formats.end(),
+                                         supported_formats.begin(), supported_formats.end());
+  if (format == swapchain_formats.end())
+  {
+    ERROR_LOG(VIDEO, "OpenXR: No usable swapchain formats.");
+    return false;
+  }
+
+  const auto selected_swapchain_format = *format;
+  m_swapchain_format = selected_swapchain_format;
+
+  DEBUG_LOG(VIDEO, "OpenXR: Using swapchain format: %d", m_swapchain_format);
+
+  // Create swapchain with texture array of size view_count.
```

Comment:
Any particular reason _why_ it is (or has to be) `view_count`, and not anything else?

Reply:
`view_count` is 2 (the number of eyeballs on a standard human), one texture layer per view. I'll make it more clear.